### PR TITLE
Add monquorum alert

### PR DESCRIPTION
**Pull request by devopsjonas**
Adds `CephMonHighNumberOfLeaderChanges`
Fixes `CephMonQuorumAtRisk` to add label selector. In the case where you have multiple clusters monitored by same Prometheus instance.

Example output:

```
"groups":
- "name": "quorum-alert.rules"
  "rules":
  - "alert": "CephMonQuorumAtRisk"
    "annotations":
      "description": "Storage cluster quorum is low. Contact Support."
      "message": "Storage quorum at risk"
      "severity_level": "error"
      "storage_type": "ceph"
    "expr": |
      count(ceph_mon_quorum_status{job="rook-ceph-mgr"} == 1) <= ((count(ceph_mon_metadata{job="rook-ceph-mgr"}) % 2) + 1)
    "for": "15m"
    "labels":
      "severity": "critical"
  - "alert": "CephMonHighNumberOfLeaderChanges"
    "annotations":
      "description": "Ceph Monitor \"{{ $labels.job }}\": instance {{ $labels.instance }} has seen {{ $value }} leader changes recently."
      "message": "Storage Cluster has seen many leader changes recently."
      "severity_level": "warning"
      "storage_type": "ceph"
    "expr": |
      rate(ceph_mon_num_elections{job="rook-ceph-mgr"}[15m]) > 3
    "for": "30m"
    "labels":
      "severity": "warning"
```

### CODE REVIEW

File: alerts/monquorum.libsonnet [hunk 1]
```diff
@@ -20,6 +20,19 @@
...
+          {
+            alert: 'CephMonHighNumberOfLeaderChanges',
+            expr: |||
+              rate(ceph_mon_num_elections{%(cephExporterSelector)s}[15m]) > 3
+            ||| % $._config,
+            'for': $._config.monQuorumLeaderChangesAlertTime,
+            labels: {
+              severity: 'warning',
+            },
+            annotations: {
```
Comment (anmolsachan) [inline alerts/monquorum.libsonnet:32]: Please update the annotations according to :

`
annotations: {
              message: '',
              description: '',
              storage_type: '',
              severity_level: '',
            }
`

**Review by anmolsachan — CHANGES_REQUESTED** [review]

File: alerts/monquorum.libsonnet [hunk 2]
```diff
@@ -7,8 +7,8 @@
           {
             alert: 'CephMonQuorumAtRisk',
             expr: |||
-              count(ceph_mon_quorum_status == 1) <= ((count(ceph_mon_metadata) %s 2) + 1)
-            ||| % '%',
+              count(ceph_mon_quorum_status{%s} == 1) <= ((count(ceph_mon_metadata{%s}) %s 2) + 1)
+            ||| % [$._config.cephExporterSelector, $._config.cephExporterSelector, '%'],
```
Comment (anmolsachan) [inline alerts/monquorum.libsonnet:11]: Is this change made assuming that there will be multiple exporter jobs working at the same time  ?

File: alerts/monquorum.libsonnet [hunk 3]
```diff
@@ -7,8 +7,8 @@
           {
             alert: 'CephMonQuorumAtRisk',
```
Comment (shtripat) [inline alerts/monquorum.libsonnet:8]: There is another similar already present at https://github.com/ceph/ceph-mixins/blob/master/alerts/monquorum.libsonnet
How its different from that?

File: alerts/monquorum.libsonnet (hunk 2 above)
Comment (devopsjonas) [reply alerts/monquorum.libsonnet:11]: Yes, I think it makes sense, as if you have 2 clusters monitored with same Prometheus the current version wouldn't work

File: alerts/monquorum.libsonnet (hunk 1 above)
Comment (devopsjonas) [reply alerts/monquorum.libsonnet:32]: I've created an issue to discuss this https://github.com/ceph/ceph-mixins/issues/54

but for now I'll refactor this to your suggestion.

**Comment by devopsjonas** [comment]
@anmolsachan fixed. PTAL :slightly_smiling_face:

File: alerts/monquorum.libsonnet (hunk 3 above)
Comment (devopsjonas) [reply alerts/monquorum.libsonnet:8]: @shtripat it's the same alert I didn't change it, except for adding a exporter selector `{%s}`

Comment (shtripat) [reply alerts/monquorum.libsonnet:8]: Ack

**Comment by umangachapagain** [comment]
looks good to me. 
Can you also run `make all` from project root and add the rule file generated?

**Comment by anmolsachan** [comment]
@devopsjonas PR looks good to me now. Please attach the generated manifest. You can refer to merged PR's for refenrence.

File: alerts/monquorum.libsonnet [hunk 4]
```diff
@@ -20,6 +20,22 @@
...
+            alert: 'CephMonHighNumberOfLeaderChanges',
+            expr: |||
+              rate(ceph_mon_num_elections{%(cephExporterSelector)s}[15m]) > 3
+            ||| % $._config,
+            'for': $._config.monQuorumLeaderChangesAlertTime,
+            labels: {
+              severity: 'warning',
+            },
+            annotations: {
+              message: 'Ceph Monitor has seen many leader changes recently.',
```
Comment (shtripat) [inline alerts/monquorum.libsonnet]: Lets avoid `Ceph` keyword in message. May be something like `Storage cluster has seen many leader changes recently`.

Comment (devopsjonas) [reply alerts/monquorum.libsonnet]: :+1: fixed :)

**Comment by devopsjonas** [comment]
Generated manifest:

```
"groups":
- "name": "quorum-alert.rules"
  "rules":
  - "alert": "CephMonQuorumAtRisk"
    "annotations":
      "description": "Storage cluster quorum is low. Contact Support."
      "message": "Storage quorum at risk"
      "severity_level": "error"
      "storage_type": "ceph"
    "expr": |
      count(ceph_mon_quorum_status{job="rook-ceph-mgr"} == 1) <= ((count(ceph_mon_metadata{job="rook-ceph-mgr"}) % 2) + 1)
    "for": "15m"
    "labels":
      "severity": "critical"
  - "alert": "CephMonHighNumberOfLeaderChanges"
    "annotations":
      "description": "Ceph Monitor \"{{ $labels.job }}\": instance {{ $labels.instance }} has seen {{ $value }} leader changes recently."
      "message": "Storage Cluster has seen many leader changes recently."
      "severity_level": "warning"
      "storage_type": "ceph"
    "expr": |
      rate(ceph_mon_num_elections{job="rook-ceph-mgr"}[15m]) > 3
    "for": "30m"
    "labels":
      "severity": "warning"
```

**Comment by devopsjonas** [comment]
@umangachapagain I've run:

```
make all            
```                                                                          
```
jsonnet -S lib/alerts.jsonnet > prometheus_alert_rules.yaml
find . -name 'vendor' -prune -o -name '*.libsonnet' -print -o -name '*.jsonnet' -print | \                                             
        while read f; do \
                jsonnet fmt -n 2 --max-blank-lines 2 --string-style s --comment-style s "$f" | diff -u "$f" -; \                       
        done
promtool check rules prometheus_alert_rules.yaml
Checking prometheus_alert_rules.yaml
  SUCCESS: 11 rules found
```
```
git status
```

```
On branch add-monquorum-alert
Your branch is up to date with 'origin/add-monquorum-alert'.

nothing to commit, working tree clean
```

I believe `.gitignore` is making sure that we don't commit generated alerts.

**Comment by shtripat** [comment]
@devopsjonas You can run `build.sh` from `extras` dir. That will generate a rule file in `manifests` dir. That file you can check in.

**Comment by devopsjonas** [comment]
Done, PTAL :)

**Review by shtripat — APPROVED** [review]
LGTM

**Comment by shtripat** [comment]
@devopsjonas we have got unit test mechanism recently added to the project. I would expect you to add unit tests for the alerting rules as part of these PRs. Otherwise both the PRs looks good to be merged.
Feel free to send another PR for unit tests separately.

**Comment by devopsjonas** [comment]
@shtripat sure will do :+1:

**Comment by shtripat** [comment]
Merging this now. Add another PR for unit tests.

**Pull request merged**